### ISSUE TITLE
Unexpected out-of-boundary behavior in `grid_sample`

### ISSUE THREAD

**Issue by turtleizzy** [pytorch/pytorch#149013]
### 🐛 Describe the bug

`grid_sample` should return padding value (0) when grid coordinates are outside `[-1, 1]`, but it does not.

I spotted this problem when I couldn't replicate the result of `grid_sample` with other libraries like `scipy.ndimage.map_coordinates` and `itk.DisplacementFieldTransform`. I experimented with labelmaps and the output segmentation from grid_sample was consistently larger than the output from other implementations.

```python
import torch

img = torch.ones([1, 1, 10, 10, 10])
grid = torch.ones([1, 1, 1, 10, 3]) + torch.linspace(-0.05, 0.05, 10)[None, None, None, :, None]

print(grid)

# tensor([[[[[0.9500, 0.9500, 0.9500],
#            [0.9611, 0.9611, 0.9611],
#            [0.9722, 0.9722, 0.9722],
#            [0.9833, 0.9833, 0.9833],
#            [0.9944, 0.9944, 0.9944],
#            [1.0056, 1.0056, 1.0056],
#            [1.0167, 1.0167, 1.0167],
#            [1.0278, 1.0278, 1.0278],
#            [1.0389, 1.0389, 1.0389],
#            [1.0500, 1.0500, 1.0500]]]]])

print(torch.nn.functional.grid_sample(img, grid, align_corners=True, mode='nearest', padding_mode='zeros'))
# tensor([[[[[1., 1., 1., 1., 1., 1., 1., 1., 1., 1.]]]]])
```

### Versions

PyTorch version: 2.1.1+cu118     
Is debug build: False          
CUDA used to build PyTorch: 11.8
ROCM used to build PyTorch: N/A                              
                                                                          
OS: Ubuntu 18.04.3 LTS (x86_64)
GCC version: (Ubuntu 7.5.0-3ubuntu1~18.04) 7.5.0
Clang version: Could not collect
CMake version: version 3.25.2
Libc version: glibc-2.27                                     
                           
Python version: 3.11.5 (main, Sep 11 2023, 13:54:46) [GCC 11.2.0] (64-bit runtime)

**Comment by turtleizzy** [comment]
Further tests revealed that the out-of-bound behavior only seemed to work when `floor(grid_location * (image_shape - 1)) > (image_shape - 1)`.

```
import torch

img = torch.ones([1, 1, 100, 100, 100])
grid = torch.ones([1, 1, 1, 10, 3]) + torch.linspace(0, 0.02, 10)[None, None, None, :, None]

print(grid)

# tensor([[[[[1.0000, 1.0000, 1.0000],
#            [1.0022, 1.0022, 1.0022],
#            [1.0044, 1.0044, 1.0044],
#            [1.0067, 1.0067, 1.0067],
#            [1.0089, 1.0089, 1.0089],
#            [1.0111, 1.0111, 1.0111],
#            [1.0133, 1.0133, 1.0133],
#            [1.0156, 1.0156, 1.0156],
#            [1.0178, 1.0178, 1.0178],
#            [1.0200, 1.0200, 1.0200]]]]])

print(torch.nn.functional.grid_sample(img, grid, align_corners=True, mode='nearest', padding_mode='zeros'))

# tensor([[[[[1., 1., 1., 1., 1., 0., 0., 0., 0., 0.]]]]])

print(grid * (torch.tensor(img.shape[2:]) - 1))

# tensor([[[[[ 99.0000,  99.0000,  99.0000],
#            [ 99.2200,  99.2200,  99.2200],
#            [ 99.4400,  99.4400,  99.4400],
#            [ 99.6600,  99.6600,  99.6600],
#            [ 99.8800,  99.8800,  99.8800],
#            [100.1000, 100.1000, 100.1000],
#            [100.3200, 100.3200, 100.3200],
#            [100.5400, 100.5400, 100.5400],
#            [100.7600, 100.7600, 100.7600],
#            [100.9800, 100.9800, 100.9800]]]]])

```

**Comment by golkir** [comment]
Upon reading the code of [grid_sampler_3d_cpu_impl ](https://github.com/pytorch/pytorch/blob/main/aten/src/ATen/native/GridSampler.cpp)
It does not seem  that the out-of-bound grid locations are zeroed explicitly. According to [documentation](https://pytorch.org/docs/stable/generated/torch.nn.functional.grid_sample.html), 

> If grid has values outside the range of [-1, 1], the corresponding outputs are handled as defined by padding_mode .... padding_mode="zeros": use 0 for out-of-bound grid locations.

However, [in the code below](https://github.com/pytorch/pytorch/blob/b90698f5ba2610dffaf33a176b50240578bfeb94/aten/src/ATen/native/GridSampler.cpp#L86C9-L97C94) we see that after getting grid locations (`ix`, `iy`, and `iz`) they are immediately passed to the `grid_sampler_compute_source_index`, which maps them to input indices before interpolation.

```cpp
for (const auto d : c10::irange(out_D)) {
          for (const auto h : c10::irange(out_H)) {
            for (const auto w : c10::irange(out_W)) {
              // get the corresponding input x, y, z co-ordinates from grid
              const scalar_t *grid_ptr_NDHW = grid_ptr_N + d * grid_sD + h * grid_sH + w * grid_sW;
              scalar_t ix = *grid_ptr_NDHW;
              scalar_t iy = grid_ptr_NDHW[grid_sCoor];
              scalar_t iz = grid_ptr_NDHW[2 * grid_sCoor];

              ix = grid_sampler_compute_source_index(ix, inp_W, padding_mode, align_corners);
              iy = grid_sampler_compute_source_index(iy, inp_H, padding_mode, align_corners);
              iz = grid_sampler_compute_source_index(iz, inp_D, padding_mode, align_corners);
```
So at this point ix, iy, iz are input indices and out-of-bounds grid values were not zeroed. What happens next [is interpolation](https://github.com/pytorch/pytorch/blob/b90698f5ba2610dffaf33a176b50240578bfeb94/aten/src/ATen/native/GridSampler.cpp#L179C17-L193C18) (in the OP case using "nearest" mode):
```cpp
...
else if (interpolation_mode == GridSamplerInterpolation::Nearest) {
                int64_t ix_nearest = static_cast<int64_t>(std::nearbyint(ix));
                int64_t iy_nearest = static_cast<int64_t>(std::nearbyint(iy));
                int64_t iz_nearest = static_cast<int64_t>(std::nearbyint(iz));

                // assign nearest neighbour pixel value to output pixel
                scalar_t *out_ptr_NCDHW = out_ptr + n * out_sN + d * out_sD + h * out_sH + w * out_sW;
                const scalar_t *inp_ptr_NC = inp_ptr_N;
                for (int64_t c = 0; c < C; ++c, out_ptr_NCDHW += out_sC, inp_ptr_NC += inp_sC) {
                  if (within_bounds_3d(iz_nearest, iy_nearest, ix_nearest, inp_D, inp_H, inp_W)) {
                    *out_ptr_NCDHW = inp_ptr_NC[iz_nearest * inp_sD + iy_nearest * inp_sH + ix_nearest * inp_sW];
                  } else {
                    *out_ptr_NCDHW = static_cast<scalar_t>(0);
                  }
                }
```
Here we can see that only if the nearest index of the source is within bounds we actually sample, otherwise we set output to zero: `*out_ptr_NCDHW = static_cast<scalar_t>(0);`. However, at this stage it only implicitly depends on the original grid coordinate value. 
Some output produced with the OP data when testing and adding some "print" statements inside [`GridSampler.cpp` code](https://github.com/pytorch/pytorch/blob/main/aten/src/ATen/native/GridSampler.cpp):
```python
print(grid)
tensor([[[[[1.0000, 1.0000, 1.0000],
           [1.0022, 1.0022, 1.0022],
           [1.0044, 1.0044, 1.0044],
           [1.0067, 1.0067, 1.0067],
           [1.0089, 1.0089, 1.0089],
           [1.0111, 1.0111, 1.0111],
           [1.0133, 1.0133, 1.0133],
           [1.0156, 1.0156, 1.0156],
           [1.0178, 1.0178, 1.0178],
           [1.0200, 1.0200, 1.0200]]]]])
print(torch.nn.functional.grid_sample(img, grid, align_corners=True, mode='nearest', padding_mode='zeros'))
....
ix before compute_source_index: 1.01111
iy before compute_source_index: 1.01111
iz before compute_source_index: 1.01111
ix source after compute_source_index: 99.55
iy source compute_source_index: 99.55
iz source compute_source_index: 99.55
ix_nearest: 100, iy_nearest: 100, iz_nearest: 100
out_ptr_NCDHW: 0
.....
tensor([[[[[1., 1., 1., 1., 1., 0., 0., 0., 0., 0.]]]]])
```
Output is zeroed after rounding to 100 (out-of-bounds index) here.

Potential solution (if following documentation as I understand it) (starting from [L94](https://github.com/pytorch/pytorch/blob/b90698f5ba2610dffaf33a176b50240578bfeb94/aten/src/ATen/native/GridSampler.cpp#L94)):

```cpp
scalar_t ix = *grid_ptr_NDHW;
scalar_t iy = grid_ptr_NDHW[grid_sCoor];
scalar_t iz = grid_ptr_NDHW[2 * grid_sCoor];

// Check if ix, iy, or iz are out of range
bool out_of_bounds = (ix < -1 || ix > 1 || iy < -1 || iy > 1 || iz < -1 || iz > 1);
scalar_t *out_ptr_NCDHW = out_ptr + n * out_sN + d * out_sD + h * out_sH + w * out_sW;
if (out_of_bounds) {
    *out_ptr_NCDHW = static_cast<scalar_t>(0);
    continue;
}
```
This ensures that out-of-bounds values are handled immediately (when they are grid coordinates), rather than relying on interpolation logic later.
If it seems reasonable I could try to implement something like this. Tagging @ezyang @malfet @zou3519 @zdevito

**Comment by golkir** [comment]
@mikaylagawarecki Hi, I’d like to contribute on the issue, but I’m unsure if this is still considered a bug or what the current status is. Also, I’d appreciate any feedback on the implementation I proposed and if my understanding of the problem is correct. Thanks!
